src/group_theory/exponent.lean

```diff
@@ -0,0 +1,159 @@
+the lowest common multiple of the order of all elements of the group `G`.
+
+## Main definitions
+
+* `monoid.exponent_exists` is a predicate on a monoid `G` saying that there is some positive `n`
+  such that `g ^ n = 1` for all `g ∈ G`.
+* `monoid.exponent` defines the exponent of a monoid `G` as the minimal positive `n` such that
+  `g ^ n = 1` for all `g ∈ G`, by convention it is `0` if no such `n` exists.
+* `add_monoid.exponent_exists` the additive version of `monoid.exponent_exists`.
+* `add_monoid.exponent` the additive version of `monoid.exponent`.
+
+## Main results
+
+* `lcm_order_eq_exponent`: For a finite group `G`, the exponent is equal to the `lcm` of the order
+  of its elements.
+
+## TODO
+* Compute the exponent of cyclic groups.
+* Refactor the characteristic of a ring to be the exponent of its underlying additive group.
+-/
+
+universe u
+
+variables (G : Type u) [monoid G]
+
+open_locale classical
+
+namespace monoid
+
+/--A predicate on a monoid saying that there is a positive integer `n` such that `g ^ n = 1`
+  for all `g`.-/
+@[to_additive "A predicate on an additive monoid saying that there is a positive integer `n` such
+  that `n • g = 0` for all `g`."]
+def exponent_exists  := ∃ n, 0 < n ∧ ∀ g : G, g ^ n = 1
+
+/--The exponent of a group is the smallest positive integer `n` such that `g ^ n = 1` for all
+  `g ∈ G` if it exists, otherwise it is zero by convention.-/
+@[to_additive "The exponent of an additive group is the smallest positive integer `n` such that
+  `n • g = 0` for all `g ∈ G` if it exists, otherwise it is zero by convention."]
+noncomputable def exponent :=
+if h : exponent_exists G then nat.find h else 0
+
+@[to_additive exponent_nsmul_eq_zero]
+lemma pow_exponent_eq_one (g : G) : g ^ exponent G = 1 :=
+begin
+  by_cases exponent_exists G,
+  { simp_rw [exponent, dif_pos h],
+    exact (nat.find_spec h).2 g},
+  { simp_rw [exponent, dif_neg h, pow_zero] }
+end
+
+@[to_additive nsmul_eq_mod_exponent]
+lemma pow_eq_mod_exponent {n : ℕ} (g : G): g ^ n = g ^ (n % exponent G) :=
+calc g ^ n = g ^ (n % exponent G + exponent G * (n / exponent G)) : by rw [nat.mod_add_div]
+  ... = g ^ (n % exponent G) : by simp [pow_add, pow_mul, pow_exponent_eq_one]
+
+@[to_additive]
+lemma exponent_pos_of_exists (n : ℕ) (hpos : 0 < n)
+(hG : ∀ g : G, g ^ n = 1) : 0 < exponent G :=
+begin
+  have h : ∃ n, 0 < n ∧ ∀ g : G, g ^ n = 1 := ⟨n, hpos, hG⟩,
+  rw [exponent, dif_pos],
+  exact (nat.find_spec h).1,
+end
+
+@[to_additive]
+lemma exponent_min' (n : ℕ) (hpos : 0 < n) (hG : ∀ g : G, g ^ n = 1) :
+  exponent G ≤ n :=
+begin
+  rw [exponent, dif_pos],
+  { apply nat.find_min',
+    exact ⟨hpos, hG⟩ },
+  { exact ⟨n, hpos, hG⟩ },
+end
+
+@[to_additive]
+lemma exponent_min (m : ℕ) (hpos : 0 < m) (hm : m < exponent G) : ∃ g : G, g ^ m ≠ 1 :=
+begin
+  by_contradiction,
+  push_neg at h,
+  have hcon : exponent G ≤ m := exponent_min' G m hpos h,
+  linarith,
+end
+
+lemma exp_punit_eq_one : exponent punit = 1 :=
```

Comment:
I think you can prove this more generally for groups that have a `subsingleton` instance such as `unit`, `zmod 1` and maybe others.
It can also be tagged `@[simp]`.

Reply:
Thanks for the suggestion! Done.